docker/bookkeeper/entrypoint.sh

```diff
@@ -48,31 +53,131 @@ export BK_tlsTrustStoreType=JKS
-else
-  # The container does not contain any BK data, it is probably a new
-  # bookie. We will format any pre-existent data and metadata before starting
-  # the bookie to avoid potential conflicts.
-  echo "format bookie data and metadata"
-  /opt/bookkeeper/bin/bookkeeper shell bookieformat -nonInteractive -force -deleteCookie
-fi
-
-echo "start bookie"
-/opt/bookkeeper/scripts/entrypoint.sh bookie
+# Override the `set -e` defined in the top of the file temporarily. Doing this is necessary as the Bookeeper
+# `common.sh` script has logic that can return non-zero status.
+set +e
+source ${SCRIPTS_DIR}/common.sh
+set -e
+
+# Create directories for multiple ledgers and journals if specified.
+create_bookie_dirs() {
+  IFS=',' read -ra directories <<< $1
+  for i in "${directories[@]}"
+  do
+      mkdir -p $i
+      if [ "$(id -u)" = '0' ]; then
+          chown -R "${BK_USER}:${BK_USER}" $i
+      fi
+  done
+}
+
+wait_for_zookeeper() {
+    echo "Waiting for zookeeper"
+    until zk-shell --run-once "ls /" ${BK_zkServers}; do sleep 5; done
+}
+
+
+create_zk_root() {
+    if [ "x${BK_CLUSTER_ROOT_PATH}" != "x" ]; then
+        echo "Creating the zk root dir '${BK_CLUSTER_ROOT_PATH}' at '${BK_zkServers}'"
+        zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH} '' false false true" ${BK_zkServers}
+    fi
+}
+
+configure_bk() {
+    # We need to update the metadata endpoint and Bookie ID before attempting to delete the cookie
+    sed -i "s|.*metadataServiceUri=.*\$|metadataServiceUri=${BK_metadataServiceUri}|" /opt/bookkeeper/conf/bk_server.conf
+    if [ ! -z "$BK_useHostNameAsBookieID" ]; then
+      sed -i "s|.*useHostNameAsBookieID=.*\$|useHostNameAsBookieID=${BK_useHostNameAsBookieID}|" ${BK_HOME}/conf/bk_server.conf
+    fi
+}
+
+# Init the cluster if required znodes do not exist in Zookeeper.
+function init_cluster() {
+    set +e
+    zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+    if [ $? -eq 0 ]; then
+        echo "Cluster metadata already exists"
+    else
+        echo "Sleeping for a random time"
+        sleep $[ ( $RANDOM % 10 )  + 1 ]s
+
+        # Create an ephemeral zk node `bkInitLock` for use as a lock.
+
+        , initiator who this node, then do init; other initiators will wait.
+        lock=`zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH}/bkInitLock1 '' true false false" ${BK_zkServers}`
+        echo "Lock: $lock"
+
+        if [ -z "$lock" ]; then
+            echo "Bookkeeper znodes do not exist in Zookeeper. Initializing a new Bookeekeper cluster."
+            /opt/bookkeeper/bin/bookkeeper shell initnewcluster
+            if [ $? -eq 0 ]; then
+                echo "initNewCluster operation succeeded"
+            else
+                echo "initNewCluster operation failed. Please check the reason."
+                echo "Exit status of initnewcluster"
+                echo $?
+                sleep 1000
```

Comment:
we can remove this sleep.

Reply:
Sure, have done that in the latest commit I've just pushed.